src/dc_location.rs

```diff
@@ -198,28 +204,26 @@ pub fn dc_get_locations(
                 timestamp_from,
                 timestamp_to,
             ],
-            |row| unsafe {
-                let mut loc: *mut _dc_location =
-                    calloc(1, ::std::mem::size_of::<_dc_location>()) as *mut _dc_location;
-                assert!(!loc.is_null(), "allocation failed");
-
-                (*loc).location_id = row.get(0)?;
-                (*loc).latitude = row.get(1)?;
-                (*loc).longitude = row.get(2)?;
-                (*loc).accuracy = row.get(3)?;
-                (*loc).timestamp = row.get(4)?;
-                (*loc).independent = row.get(5)?;
-                (*loc).msg_id = row.get(6)?;
-                (*loc).contact_id = row.get(7)?;
-                (*loc).chat_id = row.get(8)?;
-
-                if 0 != (*loc).msg_id {
+            |row| {
+                let mut loc = _dc_location::new();
+
+                loc.location_id = row.get(0)?;
+                loc.latitude = row.get(1)?;
+                loc.longitude = row.get(2)?;
+                loc.accuracy = row.get(3)?;
+                loc.timestamp = row.get(4)?;
+                loc.independent = row.get(5)?;
+                loc.msg_id = row.get(6)?;
+                loc.contact_id = row.get(7)?;
+                loc.chat_id = row.get(8)?;
+
+                if 0 != loc.msg_id {
                     let txt: String = row.get(9)?;
                     if is_marker(&txt) {
-                        (*loc).marker = to_cstring(txt);
+                        loc.marker = Some(txt);
                     }
                 }
-                Ok(loc)
+                Ok(Box::into_raw(Box::new(loc)))
```

Comment:
I would do the boxing working as part off adding it to the array, rather than as part of this loop.

Reply:
I think this is not quite safe, we can't use `free` to free the memory anymore, as the location struct was allocated from rust. But afaik the code assumes pointers inside `dc_array` are getting freed using `libc::free`, which will fail.

Reply:
`Box::into_raw(Box::new(` is safe Rust eqivalent of `malloc`. I free this pointer with a `Box::from_raw(` in `dc_array.rs` instead of `free`.